File: Backend/project/apps/financings/clases/credit.py
from datetime import datetime, timedelta
from dateutil.relativedelta import relativedelta
import sys
import os
import json

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../..')))
from apps.customers.clases.customer import Customer
from apps.InvestmentPlan.clases.investmentPlan import InvestmentPlan



class Credit:
    contador = 0
    def __init__(self, proposito, monto, plazo, tasa_interes, forma_de_pago, frecuencia_pago, fecha_inicio, tipo_credito,  customer_id,destino_id=None, fecha_vencimiento=None):
        Credit.contador+=1
        self._id = Credit.contador
        self.__proposito = 'NADA'
        self.__monto = monto
        self.__plazo = plazo
        self.__tasa_interes = tasa_interes
        self.__forma_de_pago = forma_de_pago
        self.__frecuencia_pago = frecuencia_pago
        self.__fecha_inicio = fecha_inicio
        self.__fecha_vencimiento = fecha_vencimiento if fecha_vencimiento else self.calcular_fecha_vencimiento()
        self.__tipo_credito = tipo_credito
        self.__destino_id = destino_id
        self.__customer_id = customer_id
    
    @property
    def proposito(self):
        return self.__proposito
    
    @property
    def monto(self):
        return self.__monto
    
    @property
    def plazo(self):
        return self.__plazo
    
    @property
    def tasa_interes(self):
        tasa = float(self.__tasa_interes)

        if tasa > 1:
            return (self.__tasa_interes )/100
        return (self.__tasa_interes)

    @property
    def forma_de_pago(self):
        return self.__forma_de_pago
    
    @property
    def frecuencia_pago(self):
        return self.__frecuencia_pago

    @property
    def fecha_inicio(self):
        return self.__fecha_inicio
    
    @property
    def fecha_vencimiento(self):
        return self.__fecha_vencimiento
    
    @property
    def tipo_credito(self):
        if self.__destino_id:
            self.__tipo_credito = self.__destino_id.type_of_product_or_service

        return self.__tipo_credito
    
    @property
    def destino_id(self):
        return self.__destino_id
    
    @property
    def customer_id(self):
        return self.__customer_id
    
    @proposito.setter
    def proposito(self,value):
        self.__proposito = value
    
    @monto.setter
    def monto(self,value):
        self.__monto = value
    
    @plazo.setter
    def plazo(self, value):
        self.__plazo = value
    
    @tasa_interes.setter
    def tasa_interes(self,value):
        self.__tasa_interes = value
    
    @forma_de_pago.setter
    def forma_de_pago(self,value):
        frecuenciaPago = ['NIVELADA', 'AMORTIZACIONES A CAPITAL']
        if value in frecuenciaPago:
            self.__forma_de_pago = value
        else:
            print("Frecuencia de pago no válida")
           
    @frecuencia_pago.setter
    def frecuencia_pago(self, value):
        frecuenciaPago = ['MENSUAL', 'TRIMESTRAL', 'SEMANAL']
        if value in frecuenciaPago:
            self.__frecuencia_pago = value
        else:
            print("Frecuencia de pago no válida")

    @fecha_inicio.setter
    def fecha_inicio(self, value):
        formato = '%Y-%m-%d'
        try:
            # Asegúrate de que 'value' se convierta correctamente a 'datetime'
            self.__fecha_inicio = datetime.strptime(value, formato) if isinstance(value, str) else value
        except ValueError:
            self.__fecha_inicio = datetime.now()
    
    @fecha_vencimiento.setter
    def fecha_vencimiento(self,value):
        self.__fecha_vencimiento = value
    
    @tipo_credito.setter
    def tipo_credito(self,value):
        tipoCredito = ['AGROPECURIO Y/O PRODUCTIVO','COMERCIO','SERVICIOS','CONSUMO','VIVIENDA']
        if not value in tipoCredito:
            print('Error de tipo de credito')

        self.__tipo_credito = value
    
    @destino_id.setter
    def destino_id(self,value):
        self.__destino_id = value
    
    @customer_id.setter
    def customer_id(self,value):
        self.__customer_id = value

    def calcular_fecha_vencimiento(self):
        # Convertir fecha_inicio a un objeto datetime
        fecha_inicio = datetime.strptime(self.__fecha_inicio, '%Y-%m-%d')
        plazo = self.__plazo
        # Usar relativedelta para sumar meses al objeto datetime
        fecha_vencimiento = fecha_inicio + relativedelta(months=plazo)
        # Devolver la fecha en formato string
        return fecha_vencimiento.strftime('%Y-%m-%d')
    
    def toJson(self):
        js = {
            'proposito':self.__proposito,
            'monto':self.__monto,
            'plazo':self.__plazo,
            'tasa_interes':self.__tasa_interes,
            'forma_de_pago':self.__forma_de_pago,
            'frecuecia_pago':self.__frecuencia_pago,
            'fecha_inicio':self.__fecha_inicio,
            'fecha_vencimiento':self.__fecha_vencimiento,
            'tipo_credito':self.__tipo_credito,
            'destino':self.__destino_id,
            'cliente':self.__customer_id,

        }
        return json.dumps(js, indent=4, ensure_ascii=False)
    
    def __str__(self):
        return f'Credito: \nID:{self._id},\nFecha Inicio: {self.fecha_inicio},\nPlazo: {self.__plazo} meses,\nTasa de Interes: {self.__tasa_interes} %\nFecha de Vencimiento: {self.__fecha_vencimiento},\n'

if __name__ == '__main__':
    fiador = Customer('Juan','Lopez','lopez@gmail.com','DPI','323846682','1106369','42256694','RESIDENTE','Aprobado','MASCULINO','AGRONOMO','GUATEMALTECA','COBAN','14-03-1995','SOLTERO','Indivicual (PI)')
    cliente = Customer('Juan','Lopez','lopez@gmail.com','DPI','323846682','1106369','42256694','RESIDENTE','Aprobado','MASCULINO','AGRONOMO','GUATEMALTECA','COBAN','14-03-1995','SOLTERO','Indivicual (PI)') 
    destino = InvestmentPlan('CONSUMO',1500,750,100,cliente)
    credito = Credit(destino.type_of_product_or_service,destino.total_value_of_the_product_or_service,12,0.05,'NIVELADA','MENSUAL','2024-07-14','CONSUMO',destino,fiador)
    print(credito)

File: Backend/project/staticfiles/js/class/credit.js
export class Credit {
    static contador = 0;
    #id;
    #proposito;
    #monto;
    #plazo;
    #tasa_interes;
    #forma_de_pago;
    #frecuencia_pago;
    #fecha_inicio;
    #fecha_vencimiento;
    #tipo_credito;
    #destino_id;
    #customer_id;

    constructor(proposito = '', monto = '', plazo = '', tasa_interes = '', forma_de_pago = '', frecuencia_pago = '', fecha_inicio = '', tipo_credito = '', destino_id = null, customer_id = '', fecha_vencimiento = null) {
        Credit.contador++;
        this.#id = Credit.contador;
        this.#proposito = proposito;
        this.#monto = parseFloat(monto);
        this.#plazo = parseInt(plazo);
        this.#tasa_interes = parseFloat(tasa_interes);
        this.#forma_de_pago = forma_de_pago;
        this.#frecuencia_pago = frecuencia_pago;
        this.#fecha_inicio = new Date(fecha_inicio);
        this.#fecha_vencimiento = fecha_vencimiento ? new Date(fecha_vencimiento) : this.calcularFechaVencimiento();
        this.#tipo_credito = tipo_credito;
        this.#destino_id = destino_id;
        this.#customer_id = customer_id;
    }

    get proposito() {
        return this.#proposito;
    }

    get monto() {
        return this.#monto;
    }

    get plazo() {
        return this.#plazo;
    }

    get tasaInteres() {
        const tasa = this.#tasa_interes;

        return tasa > 1 ? (tasa / 12)/100 : tasa/12;
    }

    get formaDePago() {
        return this.#forma_de_pago;
    }

    get frecuenciaPago() {
        return this.#frecuencia_pago;
    }

    get fechaInicio() {
        return this.#fecha_inicio;
    }

    get fechaVencimiento() {
        return this.#fecha_vencimiento;
        
    }

    get tipoCredito() {
        if (this.#destino_id) {
            this.#tipo_credito = this.#destino_id.type_of_product_or_service;
        }
        return this.#tipo_credito;
    }

    get destinoId() {
        return this.#destino_id;
    }

    get customerId() {
        return this.#customer_id;
    }

    set proposito(value) {
        this.#proposito = value;
    }

    set monto(value) {
        this.#monto = parseFloat(value);
    }

    set plazo(value) {
        this.#plazo = parseInt(value);
    }

    set tasaInteres(value) {
        this.#tasa_interes = parseFloat(value);
    }

    set formaDePago(value) {
        const formasDePagoValidas = ['NIVELADA', 'AMORTIZACIONES A CAPITAL'];
        if (formasDePagoValidas.includes(value)) {
            this.#forma_de_pago = value;
        } else {
            console.error("Forma de pago no válida");
        }
    }

    set frecuenciaPago(value) {
        const frecuenciasPagoValidas = ['MENSUAL', 'TRIMESTRAL', 'SEMANAL'];
        if (frecuenciasPagoValidas.includes(value)) {
            this.#frecuencia_pago = value;
        } else {
            console.error("Frecuencia de pago no válida");
        }
    }

    set fechaInicio(value) {
        this.#fecha_inicio = new Date(value);
    }

    set fechaVencimiento(value) {
        this.#fecha_vencimiento = new Date(value);
    }

    set tipoCredito(value) {
        const tiposCreditoValidos = ['AGROPECUARIO Y/O PRODUCTIVO', 'COMERCIO', 'SERVICIOS', 'CONSUMO', 'VIVIENDA'];
        if (tiposCreditoValidos.includes(value)) {
            this.#tipo_credito = value;
        } else {
            console.error('Error de tipo de crédito');
        }
    }

    set destinoId(value) {
        this.#destino_id = value;
    }

    set customerId(value) {
        this.#customer_id = value;
    }

    calcularFechaVencimiento() {
        const fechaInicio = this.#fecha_inicio;
        const fechaVencimiento = new Date(fechaInicio);
        fechaVencimiento.setMonth(fechaVencimiento.getMonth() + this.#plazo);
        return fechaVencimiento;
    }

    toJSON() {
        return {
            proposito: this.#proposito,
            monto: this.#monto,
            plazo: this.#plazo,
            tasa_interes: this.#tasa_interes,
            forma_de_pago: this.#forma_de_pago,
            frecuencia_pago: this.#frecuencia_pago,
            fecha_inicio: this.#fecha_inicio,
            fecha_vencimiento: this.#fecha_vencimiento,
            tipo_credito: this.#tipo_credito,
            destino_id: this.#destino_id,
            customer_id: this.#customer_id,
        };
    }

    toString() {
        return `Credito: \nID: ${this.#id},\nFecha Inicio: ${this.fechaInicio.toISOString().split('T')[0]},\nPlazo: ${this.#plazo} meses,\nFecha de Vencimiento: ${this.fechaVencimiento.toISOString().split('T')[0]},\nFiador: ${this.#customer_id.nombre} ${this.#customer_id.apellido}`;
    }
}
